source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,330 @@
+#include "common/buffer/buffer_impl.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace PostgresProxy {
+
+/*
+ * Postgres messages are described in official Postgres documentation:
+ * https://www.postgresql.org/docs/12/protocol-message-formats.html
+ *
+ * Most of messages start with 1-byte message identifier followed by 4-bytes length field. Few
+ * messages are defined without starting 1-byte character and are used during well-defined initial
+ * stage of connection process.
+ *
+ * Messages are composed from various fields: 8, 16, 32-bit integers, String, Arrays, etc.
+ *
+ * Structures defined below have the same naming as types used in official Postgres documentation.
+ *
+ * Each structure has the following methods:
+ * read - to read number of bytes from received buffer. The number of bytes depends on structure
+ * type. toString - method returns displayable representation of the structure value.
+ *
+ */
+
+// Template for integer types.
+// Size of integer types is fixed and depends on the type of integer.
+template <typename T> class Int {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
```

Comment:
I see the pattern of having `read` method to initialize classes in this file. Do you think making it an explicit constructor makes sense?

Reply:
I prefer to leave it as it is. There are few reasons:
- `read` may fail and return an error. Returning errors in methods is easier than from constructors.
- variadic template `Sequence` chains `read`s. Implementing similar logic using constructors would be way harder .

Reply:
Thanks!

Reply:
add doxygen comment for this method, maybe descriibing what happens when there's an error?

Reply:
thanks!